public/app/plugins/datasource/grafana-pyroscope-datasource/datasource.test.ts

```diff
@@ -105,14 +96,38 @@ describe('Pyroscope data source', () => {
     });
 
     it('should update labelSelector if there are template variables', () => {
-      ds = new PyroscopeDataSource(defaultSettings, templateSrv);
+      const ds = setupDatasource();
       const query = ds.applyTemplateVariables(
         defaultQuery({ labelSelector: `{$var="$var"}`, profileTypeId: '$var' }),
         {}
       );
       expect(query).toMatchObject({ labelSelector: `{interpolated="interpolated"}`, profileTypeId: 'interpolated' });
     });
   });
+
+  it('implements ad hoc variable support for keys', async () => {
+    const ds = setupDatasource();
+    jest.spyOn(ds, 'getResource').mockImplementationOnce(async (cb) => ['foo', 'bar', 'baz']);
+    const keys = await ds.getTagKeys({
+      filters: [],
+      timeRange: makeTimeRange('2024-01-01T00:00:00', '2024-01-01T01:00:00'),
+    });
+    expect(keys).toEqual(['foo', 'bar', 'baz'].map((v) => ({ text: v })));
+  });
+
+  it('implements ad hoc variable support for values', async () => {
+    const ds = setupDatasource();
+    jest.spyOn(ds, 'getResource').mockImplementationOnce(async (path, params) => {
+      expect(params?.label).toEqual('foo');
```

Comment:
Is there value in having an `expect` in the mock?

Reply:
Just wanted to make sure the code sends the right param to the function. It's mocked because the implementation of the getResource is not specific to this data source but what we send to that function is an implementation of the data source.